hs_core/models.py

```diff
@@ -1469,6 +1469,181 @@ def supports_folders(self):
+
+        This sets the public flag (self.raccess.public) for a resource based
+        upon application logic. It is part of AbstractResource because its result depends
+        upon resource state, and not just access control.
+
+        * This flag can only be set to True if the resource passes basic validations
+          `has_required_metata` and `has_required_content_files`
+        * setting `public` to `True` also sets `discoverable` to `True`
+        * setting `public` to `False` does not change `discoverable`
+        * setting `public` to either also modifies the AVU isPublic for the resource.
+
+        Thus, the setting public=True, discoverable=False is disallowed.
+
+        If `user` is None, access control is not checked.  This happens when a resource has been
+        invalidated outside of the control of a specific user. In this case, user can be None
+        """
+        # avoid import loop
+        from hs_core.views.utils import run_script_to_update_hyrax_input_files
+
+        # access control is separate from validation logic
+        if user is not None and not user.uaccess.can_change_resource_flags(self):
+            raise ValidationError("You don't have permission to change resource sharing status")
+
+        old_value = self.raccess.public  # is this a change?
+
+        # check that there is sufficient resource content
+        has_metadata = self.has_required_metadata
+        has_files = self.has_required_content_files
+        if value and (not has_metadata or not has_files):
+
+            if not has_metadata and not has_files:
+                msg = "Resource does not have sufficient metadata and content files to be public"
+                raise ValidationError(msg)
+
+            elif not has_metadata:
+                msg = "Resource does not have sufficient metadata to be public"
+                raise ValidationError(msg)
+
+            elif not has_files:
+                msg = "Resource does not have sufficient content files to be public"
+                raise ValidationError(msg)
+
+        else:  # make valid state change
+            self.raccess.public = value
+            if value:  # can't be public without being discoverable
+                self.raccess.discoverable = value
+            self.raccess.save()
+
+            # public changed state: set isPublic metadata AVU accordingly
+            if value != old_value:
+                self.setAVU("isPublic", self.raccess.public)
+
+                # TODO: why does this only run when something becomes public?
+                # TODO: Should it be run when a NetcdfResource becomes private?
+                # run script to update hyrax input files when private netCDF resource changes state
+                if value and settings.RUN_HYRAX_UPDATE and self.resource_type == 'NetcdfResource':
+                    run_script_to_update_hyrax_input_files(self.short_id)
```

Comment:
@alvacouch The original thinking to only run the hyrax update script when a resource becomes public is that when a resource becomes private, hyrax options will be disabled from resource landing page so a user will not be able to get to hyrax for a private resource, and the script for general clean up is run nightly as a cron job to make sure things are in sync between hydroshare irods zones and hyrax file symbolic links. Doing it this way is for performance considerations since running script will take a little bit of time. I can see the benefit of running script to update flags when a resource becomes private as well so that we don't have to rely on nightly run script for cleaning up, so if you think it is better to change it to run update script whenever a resource public status is changed, I am OK with it.